src/sidebar/util/oauth-client.ts

```diff
@@ -5,58 +5,122 @@ import { fetchJSON } from './fetch';
- * @prop {string} refreshToken - The refresh token that can be used to
- *                               get a new access token.
+ * See https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2
+ */
+type AuthorizationCodeResponse = {
+  code: string;
+  state?: string;
+
+  /** Non-standard field to differentiate this from an unsuccessful response. */
+  type: 'authorization_response';
+};
+
+/**
+ * Response sent back from popup window if the user closes the popup before
+ * completing authorization.
+ */
+type AuthorizationCanceledResponse = {
+  state?: string;
+  type: 'authorization_canceled';
+};
+
+/**
+ * Response from the authorization popup window.
  */
+type AuthorizationResponse =
+  | AuthorizationCodeResponse
+  | AuthorizationCanceledResponse;
+
+/**
+ * Access token and associated metadata received from the Hypothesis server.
+ */
+export type TokenInfo = {
+  /**
+   * The access token that should be passed to the API in an `Authorization: Bearer ${TOKEN}`
+   * header.
+   */
+  accessToken: string;
+
+  /**
+   * The date when the timestamp will expire.
+   *
+   * This uses the same epoch and units as {@link Date.now}.
+   */
+  expiresAt: number;
+
+  /** The refresh token that can be used to get a new access token. */
+  refreshToken: string;
+};
 
 /**
  * Error thrown if fetching or revoking an access token fails.
  */
 export class TokenError extends Error {
   /**
-   * @param {string} message
-   * @param {Error} cause - The error which caused the token fetch to fail
+   * Original cause of this error.
+   *
+   * See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause.
    */
-  constructor(message, cause) {
+  cause: Error;
+
+  /**
+   * @param cause - The error which caused the token fetch to fail
+   */
+  constructor(message: string, cause: Error) {
     super(message);
     this.cause = cause;
   }
 }
 
 /**
  * OAuthClient configuration.
- *
- * @typedef Config
- * @prop {string} clientId - OAuth client ID
- * @prop {string} tokenEndpoint - OAuth token exchange/refresh endpoint
- * @prop {string} authorizationEndpoint - OAuth authorization endpoint
- * @prop {string} revokeEndpoint - RFC 7009 token revocation endpoint
  */
+export type Config = {
+  /** OAuth client ID */
+  clientId: string;
+
+  /**
+   * OAuth token exchange / refresh endpoint.
+   *
+   * See https://datatracker.ietf.org/doc/html/rfc6749#section-3.2.
+   */
+  tokenEndpoint: string;
+
+  /**
+   * OAuth authorization endpoint.
+   *
+   * See https://datatracker.ietf.org/doc/html/rfc6749#section-3.1.
+   */
+  authorizationEndpoint: string;
+
+  /** RFC 7009 token revocation endpoint. */
+  revokeEndpoint: string;
+};
 
 /**
  * OAuthClient handles interaction with the annotation service's OAuth
  * endpoints.
  */
 export class OAuthClient {
+  clientId: string;
+  tokenEndpoint: string;
+  authorizationEndpoint: string;
+  revokeEndpoint: string;
```

Comment:
Not specific to this PR, but something that has popped my mind lately.

Would it make sense to set public properties in classes as `readonly`? Assuming we don't want to be able to update them, of course, but when actually having a class, I would expect updates to happen via meaningful methods, if anything, and access to properties only with read purposes.

Reply:
> Would it make sense to set public properties in classes as readonly?

Hmm... I suppose there is a tradeoff between the boilerplate of `readonly` markers / getters for preventing re-assignment statically / dynamically vs the benefit of enforcing invariants. It will probably make sense to approach this differently depending on whether we are talking about a public API exposed to third parties vs. an internal API where we are only concerned about our own accidental misuse.

In this context the intent is that all the values are provided at construction time, although it should be safe to re-assign afterwards.